plugin/storage/cassandra/spanstore/reader.go

```diff
@@ -177,26 +192,147 @@ func (s *SpanReader) FindTraces(queryParams *spanstore.TraceQueryParameters) ([]
+			}
+			return s.intersectTraceIDSlices(traceIds, tagTraceIds), nil
+		}
+		return traceIds, err
+	}
+	if len(traceQuery.Tags) > 0 {
+		return s.queryByTagsAndLogs(traceQuery.ServiceName, traceQuery.StartTimeMin, traceQuery.StartTimeMax, traceQuery.Tags, traceQuery.NumTraces)
+	}
+	return s.queryTraceIdsByService(traceQuery.ServiceName, traceQuery.StartTimeMin, traceQuery.StartTimeMax, traceQuery.NumTraces)
+}
+
+func (s *SpanReader) intersectTraceIDSlices(one, other []dbmodel.TraceID) []dbmodel.TraceID {
+	oneAsMap := dbmodel.UniqueTraceIDs{}
+	for _, i := range one {
+		oneAsMap[i] = struct{}{}
+	}
+	var retMe []dbmodel.TraceID
+	for _, i := range other {
+		if _, ok := oneAsMap[i]; ok {
+			retMe = append(retMe, i)
+		}
+	}
+	return retMe
+}
+
+type traceIDResult struct {
+	traceIDs []dbmodel.TraceID
+	err      error
+}
+
+func (s *SpanReader) queryByTagsAndLogs(serviceName string, startTime, endTime time.Time, tags map[string]string, limit int) ([]dbmodel.TraceID, error) {
+	var results []traceIDResult
+	for k, v := range tags {
+		// service_name = ? AND tag_key = ? AND tag_value = ? and start_time > ? and start_time < ? limit ?`
+		query := s.session.Query(
+			queryByTag,
+			serviceName,
+			k,
+			v,
+			model.TimeAsEpochMicroseconds(startTime),
+			model.TimeAsEpochMicroseconds(endTime),
+			limit,
+		)
+		t, err := s.executeQuery(query, s.queryTagIndexTableMetrics)
+		results = append(results, traceIDResult{err: err, traceIDs: t})
+	}
+	unionedTraceIDs := map[dbmodel.TraceID]bool{}
+	for _, result := range results {
+		if result.err != nil {
+			return nil, result.err
+		}
+		for _, traceID := range result.traceIDs {
+			unionedTraceIDs[traceID] = true
+		}
+	}
+	var retMe []dbmodel.TraceID
+	for k := range unionedTraceIDs {
+		retMe = append(retMe, k)
+	}
+	return retMe, nil
+}
+
+func (s *SpanReader) queryTraceIdsByDuration(traceQuery *spanstore.TraceQueryParameters) ([]dbmodel.TraceID, error) {
+	var traceIds []dbmodel.TraceID
+
+	minDurationMicros := (traceQuery.DurationMin.Nanoseconds() / int64(time.Microsecond/time.Nanosecond))
+	maxDurationMicros := (traceQuery.DurationMax.Nanoseconds() / int64(time.Microsecond/time.Nanosecond))
+
+	// See writer.go:indexSpanDuration  for how this is indexed
+	// This is indexed in hours since epoch, converted to seconds
+	// TODO encapsulate this calculation
+	startTimeSeconds := (traceQuery.StartTimeMin.UnixNano() / int64(time.Hour))
+	endTimeSeconds := (traceQuery.StartTimeMax.UnixNano() / int64(time.Hour))
+
+	for timeBucket := startTimeSeconds; timeBucket <= endTimeSeconds; timeBucket++ {
+		query := s.session.Query(
+			queryByDuration,
+			timeBucket,
+			traceQuery.ServiceName,
+			traceQuery.OperationName,
+			minDurationMicros,
+			maxDurationMicros,
+			traceQuery.NumTraces)
+		t, err := s.executeQuery(query, s.queryDurationIndexTableMetrics)
+		if err != nil {
+			return nil, err
+		}
+
+		traceIds = append(traceIds, t...)
+		if len(traceIds) > traceQuery.NumTraces {
```

Comment:
it would make sense to dedupe trace IDs before counting them, as the same trace night have a number of spans with similar latency.